### PR TITLE
Make split mesh work with mesh meta data without re-run mesh generators

### LINKED ISSUE #16192: Mesh generators are still get executed with --use-split
## Reason
When we have `--use-split` in the command-line, we want to just use the split mesh previously generated. This is especially important for pre-split large meshes. But now even we have `--use-split`, looks like the mesh generators are still get executed.

## Design
If I use `FileMesh` instead of `FileMeshGenerator`, `--use-split` is working as intended. But this is probably not our design.
All the mesh meta data should be in the split mesh and should be loaded properly with `--use-split`, thus we do not have to execute mesh generators at all.

## Impact
To fix the `--use-split` with mesh generators.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #16192.

### CODE REVIEW

**Comment by YaqiWang** [comment]
I think in the future we should add a `FileMeshGenerator`, which can load a checkpoint file with mesh meta data, in `ExecuteMeshGenerators` and erase all existing mesh generators for loading a split mesh. I can create an issue for that. This PR is not for doing that though.

**Comment by lindsayad** [comment]
I'll review this when those recover tests are passing

**Comment by YaqiWang** [comment]
Should fixed now.

**Comment by YaqiWang** [comment]
I think this is ready for review @lindsayad . The failure in module is irrelevant probably.

File: framework/include/base/MooseApp.h
Comment (lindsayad) [inline framework/include/base/MooseApp.h:550]: Returning writeable references breaks encapsulation. I know we have it in other places throughout MOOSE (and even libMesh), but I want to avoid it as much as possible. Would it be possible to make this `void` and add another argument to this function instead?

**Review by lindsayad — CHANGES_REQUESTED** [review]
This looks ok to me. You are definitely adding important capability here; thank you for that! I would like @permcody to weigh in on whether writing this `MeshMetaData` to a distinct file is the right thing to do.

**Review by lindsayad — CHANGES_REQUESTED** [review]
This looks ok to me. You are definitely adding important capability here; thank you for that! I would like @permcody to weigh in on whether writing this `MeshMetaData` to a distinct file is the right thing to do.

**Comment by YaqiWang** [comment]
The checkpoint files for restart/recover has separate file for mesh meta data as well. The initial main purpose is to skip mesh generators when using pre-split mesh, which is important for large benchmarks where executing mesh generators is extremely costly.

File: framework/src/mesh/FileMesh.C
Comment (YaqiWang) [inline framework/src/mesh/FileMesh.C:129]: This is where `setRestartableDataMap` is used. If this code is in mesh generators instead of `FileMesh`, we can load the mesh meta data locally and then let the mesh generator to push them into app with `declareMeshProperty`.

Comment (lindsayad) [reply framework/src/mesh/FileMesh.C:129]: Avoiding the new mutator would be great

Comment (lindsayad) [reply framework/src/mesh/FileMesh.C:129]: `readRestartableData` takes a `const RestartableDataMaps &` not a `RestartableDataMaps &`, so AFAICT there is no need for the new mutator. You should be able to just use the existing `getRestartableDataMap`

Comment (YaqiWang) [reply framework/src/mesh/FileMesh.C:129]: You are right, but how is this possible? Supposed we are loading the data into `meta_data`.

Comment (YaqiWang) [reply framework/src/mesh/FileMesh.C:129]: Oh they are loaded into RestartableDataValuePair which is using `std::unique_ptr` to hold the data.

**Comment by YaqiWang** [comment]
I have some thoughts for this PR:

1. move the mesh meta data in checkpoint files from something like `file_base_cp/0020_mesh.rd` to `file_base_cp/0020_mesh.cpr/mesh_meta_data.rd`. In this way `file_base_cp/0020_mesh.cpr/` contains the mesh consistent with the one generated by split.
2. move the mesh meta data loading into `FileMeshGenerator`.
3. replace `FileMesh` to `FileMeshGenerator` in the `SetupMeshAction` for split.

For Item 1, what the other meta data are? I guess I should leave the others still at the base folder.

**Comment by permcody** [comment]
Thanks for tagging me. This is an important capability so I support the PR. I do agree with Alex that the API should be cleaned up a bit for the reasons he stated. I do think we need a separate file too, but I'm not 100% sure if we need yet another new separate file without trying to remember the entire original design. We do write the metadata to a separate file already because we have to due to the early nature of the generators running first and needing to be able to read that data super early for a normal run. We can't wait around and reuse the normal backup/restore files. The only question I have is whether this in some weird edge case would end up creating two of these metadata files or not. Did you consider that case? My thinking is probably not, but I'm wondering if you thought about it.

**Comment by YaqiWang** [comment]
I do not think we can create multiple mesh meta data files in one run. The data are either written in `SplitMeshAction` or in `Checkpoint` (grep `writeRestartableData` we see only two places in these two are using it). One thing I am not clear is that `_restartable_meta_data` is a map, and mesh meta data is just one entry in this map. What else are held in this map?

**Comment by YaqiWang** [comment]
I only find this `MooseApp::registerRestartableDataMapName` is called once in MooseApp constructor for mesh meta data.

**Comment by YaqiWang** [comment]
I made check point files and split mesh files are more consistent in dc588e4, which changes the file name of mesh meta data. If users are not abusing these data, this should not have any impact to them. Now we can use `FileMeshGenerator` to load the mesh from a checkpoint file. I think this feature is pretty cool.

**Comment by lindsayad** [comment]
@YaqiWang I'm adding a commit here for skipping that porous flow test per #16255

**Comment by YaqiWang** [comment]
Do you guys have more comments?

**Review by lindsayad — DISMISSED** [review]

**Comment by lindsayad** [comment]
That Mac failure looks real

**Comment by YaqiWang** [comment]
At a clinic now, will take a look when back. Should be pretty simple to fix.

**Comment by YaqiWang** [comment]
Looking. It is strange the test is looking for `0000_mesh.cpr` where `0003_mesh.cpr` should be the `LATEST`. Need to compile my moose and be able to debug it.

**Comment by YaqiWang** [comment]
It is the logic in `MooseUtils::getLatestCheckpointFileHelper` screwing thing up. The first time we generate the checkpoint files with the steady-state run, we can run the transient restart properly. But if I re-run the steady-state, the timestamp on the check point files makes the function not returning the latest timestep. It is super strange. Do you have any clue?

**Comment by YaqiWang** [comment]
I think I need your help to understand this. The first time in the folder I run master_ss.i, I got the checkpoint files like:
```
(griffin) [wangy2][~/projects/check/griffin/moose/modules/combined/test/tests/restart-transient-from-ss-with-stateful]> ll -t master_ss_checkpoint_cp/ 
total 1072
drwxr-xr-x  11 wangy2     352 Nov 19 15:50 ./
drwxr-xr-x   2 wangy2      64 Nov 19 15:50 0000_mesh.cpr/
drwxr-xr-x   4 wangy2     128 Nov 19 15:50 0002_mesh.cpr/
-rw-r--r--   1 wangy2  254021 Nov 19 15:50 0002.rd-0
-rw-r--r--   1 wangy2     352 Nov 19 15:50 0002.xdr
-rw-r--r--   1 wangy2   11468 Nov 19 15:50 0002.xdr.0000
drwxr-xr-x   4 wangy2     128 Nov 19 15:50 0001_mesh.cpr/
-rw-r--r--   1 wangy2  254021 Nov 19 15:50 0001.rd-0
-rw-r--r--   1 wangy2     352 Nov 19 15:50 0001.xdr
-rw-r--r--   1 wangy2   11468 Nov 19 15:50 0001.xdr.0000
drwxr-xr-x  14 wangy2     448 Nov 19 15:50 ../
```
With these checkpoint files, I was able to run master_tr.i successfully. But If I then re-run master_ss.i, I got the checkpoint files re-written:
```
(griffin) [wangy2][~/projects/check/griffin/moose/modules/combined/test/tests/restart-transient-from-ss-with-stateful]> ll -t master_ss_checkpoint_cp/ 
total 1088
drwxr-xr-x  11 wangy2     352 Nov 19 15:50 ./
drwxr-xr-x   2 wangy2      64 Nov 19 15:50 0000_mesh.cpr/
-rw-r--r--   1 wangy2  254021 Nov 19 15:50 0002.rd-0
-rw-r--r--   1 wangy2     352 Nov 19 15:50 0002.xdr
-rw-r--r--   1 wangy2   11468 Nov 19 15:50 0002.xdr.0000
-rw-r--r--   1 wangy2  254021 Nov 19 15:50 0001.rd-0
-rw-r--r--   1 wangy2     352 Nov 19 15:50 0001.xdr
-rw-r--r--   1 wangy2   11468 Nov 19 15:50 0001.xdr.0000
drwxr-xr-x   4 wangy2     128 Nov 19 15:50 0002_mesh.cpr/
drwxr-xr-x   4 wangy2     128 Nov 19 15:50 0001_mesh.cpr/
drwxr-xr-x  14 wangy2     448 Nov 19 15:50 ../
```
Then I cannot run master_tr.i properly because it fails locating 0003 as the latest. What is going on?

**Comment by YaqiWang** [comment]
Maybe It’s related with file deletion which I didn’t pay attention.

**Comment by lindsayad** [comment]
I can take a look but it may not be for a few days. It is a bizarre failure for sure

**Comment by YaqiWang** [comment]
No worries I am guessing because I move mesh meta data file into the sub folder and delete it later than the mesh files, that folder is not properly deleted. Code in Checkpoint need to be shuffled a little. Away from home and will try it out soon. Just need you to keep an eye on this PR. I want it to be merged asap.

**Comment by lindsayad** [comment]
I will be watching thanks!

**Comment by lindsayad** [comment]
@YaqiWang if you push to this again, please remove my commit. @WilkAndy has fixed that issue in https://github.com/idaholab/moose/pull/16259

**Comment by arovinelli** [comment]
> are not abusing these data, this should not have any impact to them. Now we can use `FileMeshGenerator` to load the mesh from a checkpoint file. I think this feature is pretty cool.

I definitely support this

**Comment by YaqiWang** [comment]
This should pass all the tests...

**Comment by arovinelli** [comment]
would `FileMeshGenerator` work when loading a pre-split mesh without using  --use-split option?

**Comment by YaqiWang** [comment]
Yes.

**Comment by arovinelli** [comment]
> Yes.

and each task will load only its portion of the mesh?
can we add a test for this case (I did see only one with replicate mesh mode)

**Comment by YaqiWang** [comment]
You will have to use the same number of processors and have `--distributed-mesh` on command-line. Otherwise you cannot load the file. I did not check but assume if you do this, each processor will load its own part of the mesh.

**Comment by arovinelli** [comment]
> You will have to use the same number of processors and have `--distributed-mesh` on command-line. Otherwise you cannot load the file. I did not check but assume if you do this, each processor will load its own part of the mesh.

Cool. But if you think this is possible and works I would suggest adding a test. At least when we find other problems we are sure is not due to some unforeseen behavior.

**Comment by YaqiWang** [comment]
I will.

**Comment by arovinelli** [comment]
> I will.

Thanks

**Comment by YaqiWang** [comment]
Something is not working as I expected. I had a misunderstanding. Thanks for making me look into this further. The test I added is not checking this...

**Comment by YaqiWang** [comment]
Unfortunately we cannot load a binary mesh meta data file without knowing what types of data are stored there. So we cannot have `FileMeshGenerator` to load the mesh meta data directly because it does not know what data type the data file contains. It can load the data in the file only for already declared data types. When we use `--use-split`, we still need to construct mesh generators and mesh generators should declare meta data type within the constructor (I am going to add code to protect this) so that we can load the mesh meta data properly. I am not sure if loading data without pre-declaration even possible with c++. @permcody can probably make sure on this.

If you really want to load something from a checkpoint mesh file or a split mesh file, you can derive `FileMeshGenerator` and declare the data in your derived mesh generator with the same mesh generator name and the same data name. With this PR, at least the mesh meta data file is generated.

**Comment by arovinelli** [comment]
At list from my end deriving from `FileMeshGenrator` is not too bad. Though I think the test you want to skip would be a good example for all users that need to use this capability. You need to read metadata on rank zero and send the metadata to other ranks, otherwise you will risk a race condition.
Also, I would still like to see `FileMeshGenrator` properly loading pre-split mesh in distributed mode.

**Comment by arovinelli** [comment]
@YaqiWang I played a bit with your PR as I'm interested in the capability of reloading a split mesh from a meshgenerator.
I did some tests and things don't go well in debug using --use-split ot reloading cpr file in distributed mode.
In both cases I hit the following assertion:

Assertion `mesh.is_prepared()' failed.
Should be easy enough to fix.

could you please include the additional test I added here https://github.com/idaholab/moose/commit/a0de81bb13d5920cb764abce686c1602052758f9 after you fixed that assertion?
Thanks

**Comment by arovinelli** [comment]
@YaqiWang I added better tests and fixed the assertion (check if my changes are compatible with everything else as I added calls to prepare_for_use). Now I compare a serial solution to a distributed solution. All changes i added are in this branch https://github.com/arovinelli/moose/tree/mesh_meta_data. 
Now I feel more comfortable about everything running smoothly. Still, I would like to see a test/example for reloading metadata in a distributed environment running properly.

**Comment by YaqiWang** [comment]
I tried the following and did not see the assertion you saw:
```
moose/test/tests/kernels/2d_diffusion]> ../../../moose_test-dbg -i 2d_diffusion_test.i --split-mesh 4 --split-file aaa
mpirun -np 4 ../../../moose_test-dbg -i 2d_diffusion_test.i --use-split --split-file aaa
mpirun -np 4 ../../../moose_test-dbg -i 2d_diffusion_test.i Mesh/inactive=square Mesh/fmg/type=FileMeshGenerator Mesh/fmg/file=aaa.cpr --distributed-mesh
```

**Comment by arovinelli** [comment]
> I tried the following and did not see the assertion you saw:
> 
> ```
> moose/test/tests/kernels/2d_diffusion]> ../../../moose_test-dbg -i 2d_diffusion_test.i --split-mesh 4 --split-file aaa
> mpirun -np 4 ../../../moose_test-dbg -i 2d_diffusion_test.i --use-split --split-file aaa
> mpirun -np 4 ../../../moose_test-dbg -i 2d_diffusion_test.i Mesh/inactive=square Mesh/fmg/type=FileMeshGenerator Mesh/fmg/file=aaa.cpr --distributed-mesh
> ```

@YaqiWang and @lindsayad apparently the Assertion I hit it's a problem of my build.

**Comment by YaqiWang** [comment]
I added one commit to test file mesh generator for loading checkpoint files. Looks we do not have dedicated tests for this especially in the distributed mesh mode. It is not very relevant with this PR, but I think it is better to add them.

File: test/tests/meshgenerators/file_mesh_generator/tests
Comment (arovinelli) [inline test/tests/meshgenerators/file_mesh_generator/tests:43]: shouldn't we add min_parallel and max_parallel = something > 1 to make sure both tests are executed with the same number of cpu?

Comment (YaqiWang) [reply test/tests/meshgenerators/file_mesh_generator/tests:43]: no, because the parallel test target will use the same number of processors for all the test unless you specify to not to.

**Comment by YaqiWang** [comment]
@lindsayad This is ready unless you want to check it one more time.

**Comment by lindsayad** [comment]
Is @arovinelli happy?

**Comment by arovinelli** [comment]
> Is @arovinelli happy?

@lindsayad To say the truth I would like to see a practical example of reloading metadata not only their names (eg. a subclass of `FileMeshGenrator` as pointed out by @YaqiWang previously). The example will be of good value for whoever is going to use this approach. I already pointed this before and I feel without this example the PR is somehow incomplete.

**Comment by YaqiWang** [comment]
That is not this MR about. This MR is to skip execution of mesh generators when using split mesh.

**Comment by arovinelli** [comment]
> That is not this MR about. This MR is to skip execution of mesh generators when using split mesh.

@YaqiWang Then I misinterpreted your goals, but what if any object requires the saved metadata? IMHO I still think you should show that you can reload the metadata, and that those metadata can be accessed by other MOSOE objects. These are my two cents. I'll let you and @lindsayad decide.

**Comment by YaqiWang** [comment]
With this PR, when we use the split mesh, the meta data will be loaded properly. Please understand what I am doing to make comments...

**Comment by arovinelli** [comment]
> With this PR, when we use the split mesh with `--use-split`, the meta data will be loaded properly. Please understand what I am doing to make comments...

this your previous comment:

> If you really want to load meta data from a checkpoint mesh file or a split mesh file, you can derive FileMeshGenerator and declare the data in your derived mesh generator with the same mesh generator name and the same data name. With this PR, at least the mesh meta data file is generated. However, I am not going to add this test in this PR because I am not sure how useful this feature will be (at least I cannot think a use case in my applications).

**Comment by YaqiWang** [comment]
What do you want to say here? I suggested you to have a solution for your issue and added some capabilities in this PR that can be useful to you. Now it is my fault? Did you ever read this PR containing a test in `test/tests/meshgenerators/meta_data_store/tests`?

**Comment by arovinelli** [comment]
@YaqiWang I think you shouldn't take this personally, I appreciate your comment proposing a solution, and I never gave a fault to anyone. I suggested to add an example without `--use-split` so that it could be used as guidance by everyone, not only me. I also mentioned and reiterate here you and @lindsayad decide if it is appropriate to add the example or not, meaning that should not block merge. That's all.

File: framework/src/base/MooseApp.C
Comment (lindsayad) [inline framework/src/base/MooseApp.C:1709]: I believe we will never actually get to this line because of the `return` in line 1705. Seems like we should either eliminate that early return or also put `_executing_mesh_generators = false` before the early return

File: framework/src/base/MooseApp.C
Comment (lindsayad) [inline framework/src/base/MooseApp.C:2195]: Lol did no one ever use this before?

File: test/tests/meshgenerators/file_mesh_generator/tests
Comment (lindsayad) [inline test/tests/meshgenerators/file_mesh_generator/tests:43]: Should this requirement include meta data in it?

**Comment by YaqiWang** [comment]
This PR is critical to me and I do not want to delay this further. I made things clear that (load mesh meta data in a separate mesh generator) is not the goal of this PR and may not be a feature to be added in framework. I did thought that is a byproduct of this PR before but realized in a hard way that there is no way to load the data without knowing the data type with the current design of binary writing and loading. I hope you understand these.

**Review by lindsayad — CHANGES_REQUESTED** [review]
My only concern is the `_executing_mesh_generators` flag, which I'm sorry I didn't comment on before. Otherwise I think this is fine to go in as-is. If there is more work that @arovinelli wishes to happen beyond the scope of the title of this PR, then let's create a separate issue for it.

File: framework/src/base/MooseApp.C
Comment (YaqiWang) [reply framework/src/base/MooseApp.C:2195]: hehe, this is a potential bug.

File: framework/src/base/MooseApp.C
Comment (YaqiWang) [reply framework/src/base/MooseApp.C:1709]: right... although this line is not important. I will keep this line because the early return logic does not say we will always hit the early return despite that could be the reality.

File: test/tests/meshgenerators/file_mesh_generator/tests
Comment (YaqiWang) [reply test/tests/meshgenerators/file_mesh_generator/tests:43]: No because it does. We have no way currently to load the meta data in a separate mesh generator without know what data types are stored in the binary file.

**Comment by arovinelli** [comment]
> This PR is critical to me and I do not want to delay this further. I made things clear that (load mesh meta data in a separate mesh generator) is not the goal of this PR and may not be a feature to be added in framework. I did thought that is a byproduct of this PR before but realized in a hard way that there is no way to load the data without knowing the data type with the current design of binary writing and loading. I hope you understand these.

@YaqiWang I understand your need to merge this quickly. I also understand that reloading is dependent upon the dataformat.

> If there is more work that @arovinelli wishes to happen beyond the scope of the title of this PR, then let's create a separate issue for it.

@lindsayad As soon as I will have some time I will open an issue for adding an example of reloading mesh metadata via a subclass of `FileMesGenerator`. I think it will be useful.

File: framework/src/base/MooseApp.C
Comment (lindsayad) [reply framework/src/base/MooseApp.C:1712]: 👍

**Review by lindsayad — APPROVED** [review]

**Pull request merged**
